metronome/rocksdb/src/io/iohk/metronome/rocksdb/RocksDBStore.scala

```diff
@@ -0,0 +1,374 @@
+
+  private def withWriteLock[A](fa: F[A]): F[A] =
+    Sync[F].bracket(lockWrite)(_ => fa)(_ => unlockWrite)
+
+  // In case there's a write operation among the reads and we haven't
+  // taken out a write lock, we can upgrade the read lock for the duration
+  // of the write, then downgrade it back to read when we're done.
+  // See here for the rules up upgrading/downgrading:
+  // https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/ReentrantReadWriteLock.html
+  private def withLockUpgrade[A](fa: F[A]): F[A] =
+    Sync[F].bracket {
+      unlockRead >> lockWrite
+    }(_ => fa) { _ =>
+      lockRead >> unlockWrite
+    }
+
+  /** Execute the accumulated write operations in a batch. */
+  private val writeBatch: ReaderT[F, BatchEnv, Unit] =
+    ReaderT {
+      case (opts, batch) if batch.hasPut() || batch.hasDelete() =>
+        Sync[F].delay {
+          db.write(opts, batch)
+          batch.clear()
+        }
+      case _ =>
+        ().pure[F]
+    }
+
+  /** Execute one `Get` operation. */
+  private def read[K, V](op: Get[Namespace, K, V]): F[Option[V]] = {
+    for {
+      kbs <- encode(op.key)(op.keyCodec)
+      mvbs <- Sync[F].delay[Option[Array[Byte]]] {
+        Option(db.get(handles(op.namespace), kbs))
+      }
+      mv <- mvbs match {
+        case None =>
+          none.pure[F]
+
+        case Some(bytes) =>
+          decode(bytes)(op.valueCodec).map(_.some)
+      }
+    } yield mv
+  }
+
+  /** Collect writes in a batch, until we either get to the end, or there's a read.
+    * This way writes are atomic, and reads can see their effect.
+    *
+    * In the next version of RocksDB we can use transactions to make reads and writes
+    * run in isolation.
+    */
+  private val batchingCompiler: KVNamespacedOp ~> Batch =
+    new (KVNamespacedOp ~> Batch) {
+      def apply[A](fa: KVNamespacedOp[A]): Batch[A] =
+        fa match {
+          case op @ Put(n, k, v) =>
+            ReaderT { case (_, batch) =>
+              for {
+                kbs <- encode(k)(op.keyCodec)
+                vbs <- encode(v)(op.valueCodec)
+                _ = batch.put(handles(n), kbs, vbs)
+              } yield ()
+            }
+
+          case op @ Get(_, _) =>
+            // Execute any pending deletes and puts before performing the read.
+            writeBatch >> ReaderT.liftF(read(op))
+
+          case op @ Delete(n, k) =>
+            ReaderT { case (_, batch) =>
+              for {
+                kbs <- encode(k)(op.keyCodec)
+                _ = batch.delete(handles(n), kbs)
+              } yield ()
+            }
+        }
+    }
+
+  /** Intended for reads, with fallback to writes. */
+  private val nonBatchingCompiler: KVNamespacedOp ~> F =
+    new (KVNamespacedOp ~> F) {
+      def apply[A](fa: KVNamespacedOp[A]): F[A] =
+        fa match {
+          case op @ Get(_, _) =>
+            read(op)
+          case op =>
+            withLockUpgrade {
```

Comment:
My understanding is that [Free.foldMap](https://github.com/typelevel/cats/blob/master/free/src/main/scala/cats/free/Free.scala#L142-L147) is using a trampoline with `tailRecM`, so it's not recursing while holding the write lock taken out for doing the `Put` or the `Delete`. 

Instead, for this compiler, every step is transformed to a `Task` individually, with the result returned, so that the `Free.FlatMap` that wraps it can take it and pass it on to the continuation. By that time the lock will again be downgraded. 
```scala
new (KVNamespacedOp ~> F) {
      def apply[A](fa: KVNamespacedOp[A]): F[A] =
```

Reply:
thanks for explanation !:+1: